agent/src/main/java/org/openremote/agent/protocol/mqtt/AbstractMQTT_IOClient.java

```diff
@@ -95,41 +144,65 @@ protected AbstractMQTT_IOClient(String clientId, String host, int port, boolean
+                // this allows us to track any failed subscriptions on reconnect
+
+                boolean reconnect = false;
+
+                synchronized (this) {
+                    if (connectionStatus == ConnectionStatus.CONNECTED) {
+                        onConnectionStatusChanged(ConnectionStatus.CONNECTING);
+                        reconnect = true;
                     }
                 }
-                if (context.getCause() instanceof Mqtt3DisconnectException) {
-                    LOG.info("Client disconnect '" + getClientUri() + "': initiator=" + context.getSource());
-                } else if (context.getCause() instanceof Mqtt3ConnAckException) {
-                    LOG.info("Connection rejected by the broker '" + getClientUri() + "': reasonCode=" + ((Mqtt3ConnAckException)context.getCause()).getMqttMessage().getReturnCode() + ", initiator=" + context.getSource());
-                } else if (context.getCause() instanceof ConnectionClosedException) {
-                    LOG.info("Connection closed by " + context.getSource() + " '" + getClientUri() + "': initiator=" + context.getSource());
-                } else if (context.getCause() instanceof ConnectionFailedException) {
-                    LOG.log(Level.INFO, "Connection failed '" + getClientUri() + "': initiator=" + context.getSource(), context.getCause());
+
+                if (reconnect) {
+                    if (context.getCause() instanceof Mqtt3DisconnectException) {
+                        LOG.info("Connection error uri=" + getClientUri() + ", initiator=" + context.getSource());
+                    } else if (context.getCause() instanceof Mqtt3ConnAckException) {
+                        LOG.info("Connection rejected uri=" + getClientUri() + ", reasonCode=" + ((Mqtt3ConnAckException)context.getCause()).getMqttMessage().getReturnCode());
+                    } else if (context.getCause() instanceof ConnectionClosedException) {
+                        LOG.info("Connection closed uri=" + getClientUri() + ", initiator=" + context.getSource());
+                    } else if (context.getCause() instanceof ConnectionFailedException) {
+                        LOG.info("Connection failed uri=" + getClientUri() + ", initiator=" + context.getSource() + ", message=" + context.getCause().getMessage());
+                    }
+                    doReconnect();
                 }
-                onConnectionStatusChanged(ConnectionStatus.DISCONNECTED);
-            })
-            .automaticReconnect()
-            .initialDelay(RECONNECT_DELAY_INITIAL_MILLIS, TimeUnit.MILLISECONDS)
-            .maxDelay(RECONNECT_DELAY_MAX_MILLIS, TimeUnit.MILLISECONDS)
-            .applyAutomaticReconnect();
+            });
+//            .automaticReconnect()
+//            .initialDelay(RECONNECT_DELAY_INITIAL_MILLIS, TimeUnit.MILLISECONDS)
+//            .maxDelay(RECONNECT_DELAY_MAX_MILLIS, TimeUnit.MILLISECONDS)
+//            .applyAutomaticReconnect();
```

Comment:
Can the commented code be removed?

Reply:
I have an open feature request with HiveMQ MQTT client to provide callback whenever subscriptions are auto-retried on reconnection so this code could be un-commented then to reinstate the builtin reconnector, I'll remove this and leave a comment about rollback to earlier version